Web Inspector: [Cocoa] `RemoteInspector` fails to send automatic inspection candidates created on a non-main thread

patrickangle opened the pull request:
#### 2cf06cdf0b2a0ecb699830de3b55221120003e7a
<pre>
Web Inspector: [Cocoa] `RemoteInspector` fails to send automatic inspection candidates created on a non-main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=242267">https://bugs.webkit.org/show_bug.cgi?id=242267</a>
rdar://51415024

Reviewed by Devin Rousso.

Automatic inspection of JSContexts works by spinning the thread a JSContext is created on when the feature is enabled at
the device level (via webinspectord) until we can ask an attached debugger(s) if we should automatically inspect that
context. At that same time, the attached debugger will let us know if we should formally pause script execution, or if
we should allow execution to continue normally. Upon receiving the yes/no from the debugger, we unspin the thread and if
requested start the debug session/pause JS execution.

This previously only reliably worked for the main thread and for contexts created after a JSContext was created on the
main thread. This was because there is initialization that must be done on the main thread (WTF::initialize). The issues
arose because while we would schedule initialization to later occur on the main thread, non-main-threads would still
carry on checking to see if, among other things, automatic inspection was enabled. Because the device-global automatic
inspection enablement wasn&apos;t read until after main thread initialization, we always declined to automatically inspect
non-main-thread contexts created before initialization had completed on the main thread.

This tangle of dependancies is unravelled by doing a few different things in tandem:
1. Get the global automatic inspection state before scheduling initialization - `notify_get_state` is designed for
cross-process and cross-thread messaging, so there is no need for us to be on the main thread to do so.
2. Keep track of the RemoteInspector singleton pending initialization.
3. Consider a &quot;targets&quot; (called debuggables by the client) for automatic inspection and spin them if automatic
inspection is enabled (since we now get that state before this point for all threads) and only bail if we are both
disabled and not pending initialization.

As part of this, it is now much more likely that multiple JSContexts may need to be considered candidates for automatic
inspection at the same time, so we also fix an existing limitation by allowing ourselves to keep track of multiple
candidates, since we may build a small set of candidates before initialization has completed.

* Source/JavaScriptCore/inspector/remote/RemoteInspector.cpp:
(Inspector::RemoteInspector::setupFailed):
(Inspector::RemoteInspector::setupCompleted):
(Inspector::RemoteInspector::waitingForAutomaticInspection):
* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
(Inspector::RemoteInspector::singleton):
(Inspector::RemoteInspector::initialize):
(Inspector::RemoteInspector::setPendingMainThreadInitialization):
(Inspector::RemoteInspector::updateAutomaticInspectionCandidate):
(Inspector::RemoteInspector::sendAutomaticInspectionCandidateMessage):
(Inspector::RemoteInspector::start):
(Inspector::RemoteInspector::stopInternal):
(Inspector::RemoteInspector::setupXPCConnectionIfNeeded):
(Inspector::RemoteInspector::xpcConnectionFailed):
(Inspector::RemoteInspector::receivedSetupMessage):
(Inspector::RemoteInspector::receivedAutomaticInspectionConfigurationMessage):
(Inspector::RemoteInspector::receivedAutomaticInspectionRejectMessage):
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorGlib.cpp:
(Inspector::RemoteInspector::stopInternal):
(Inspector::RemoteInspector::sendAutomaticInspectionCandidateMessage):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocket.cpp:
(Inspector::RemoteInspector::stopInternal):
(Inspector::RemoteInspector::sendAutomaticInspectionCandidateMessage):

Canonical link: <a href="https://commits.webkit.org/252569@main">https://commits.webkit.org/252569@main</a>
</pre>

JosephPecoraro commented on Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
```diff
@@ -301,8 +306,7 @@ class JS_EXPORT_PRIVATE RemoteInspector final
     bool m_automaticInspectionEnabled { false };
-    bool m_automaticInspectionPaused { false };
-    TargetID m_automaticInspectionCandidateTargetIdentifier { 0 };
+    HashSet<TargetID, WTF::IntHash<unsigned>, WTF::UnsignedWithZeroKeyHashTraits<unsigned>> m_pausedAutomaticInspectionCandidates;
```
I wonder if we can use `WTF_GUARDED_BY_LOCK(m_mutex)` so the compiler can ensure this set is guarded?

JosephPecoraro commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -125,71 +123,95 @@ static bool globalAutomaticInspectionState()
-
-        auto targetIdentifier = target->targetIdentifier();
-
-        // FIXME: We should handle multiple debuggables trying to pause at the same time on different threads.
```
I forgot we had this FIXME, nice to see you're making this possible now =)

JosephPecoraro commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -125,71 +123,95 @@ static bool globalAutomaticInspectionState()
+void RemoteInspector::initialize()
+{
+    ASSERT(isMainThread());
+    if (!m_pendingMainThreadInitialization)
```
This access to `m_pendingMainThreadInitialization` does not have the `Lock`, but all other accesses do.

You could perhaps just make the field an `Atomic<bool>`?

JosephPecoraro commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -99,21 +99,19 @@ static bool globalAutomaticInspectionState()
-                // The main thread initialization is required for certain WTF values that need to be initialized
-                // on the "real" main thread. We should investigate a better way to handle this.
-                dispatch_async(dispatch_get_main_queue(), initialize);
+                // Aquire the global automatic inspection state so we can determine if we should actually be pausing
```
```suggestion
                // Acquire the global automatic inspection state so we can determine if we should actually be pausing
```

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -99,21 +99,19 @@ static bool globalAutomaticInspectionState()
-                    shared.get().start();
-            };
+            {
+                Locker locker { shared.get().m_mutex };
```
NIT: instead of all these `shared.get()`, could/should we `shared->` instead?

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -270,7 +288,7 @@ static bool globalAutomaticInspectionState()
     updateHasActiveDebugSession();
 
-    m_automaticInspectionPaused = false;
+    m_pausedAutomaticInspectionCandidates.clear();
```
Does this need to be guarded by `m_mutex`?

patrickangle replied:
`stopInternal` is only called from places where a lock is already taken.

dcrousso replied:
ah nice in that case can we `ASSERT(m_mutex.isLocked())`?

patrickangle replied:
I can do us one better for these and actually `WTF_REQUIRES_LOCK(m_mutex)` the method itself which compile-time asserts a lock is taken before calling. For now I'll do this method and all of the `received*` methods. Also I have to do that because following @JosephPecoraro's suggestion for using `WTF_GUARDED_BY_LOCK(m_mutex)` for `m_pausedAutomaticInspectionCandidates` enforces the places that touch that set signal (either using a new lock or a method annotation) that a lock is present.

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -571,8 +591,7 @@ static bool globalAutomaticInspectionState()
     if (is<RemoteInspectionTarget>(target)) {
-        bool isAutomaticInspection = m_automaticInspectionCandidateTargetIdentifier == target->targetIdentifier();
-
+        bool isAutomaticInspection = m_pausedAutomaticInspectionCandidates.contains(target->targetIdentifier());
```
Does this need to be guarded by `m_mutex`?

patrickangle replied:
`receivedSetupMessage` is only called from places where a lock is already taken.

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -730,8 +749,8 @@ static bool globalAutomaticInspectionState()
-    if (!m_automaticInspectionEnabled && m_automaticInspectionPaused)
-        m_automaticInspectionPaused = false;
+    if (!m_automaticInspectionEnabled)
+        m_pausedAutomaticInspectionCandidates.clear();
```
Does this need to be guarded by `m_mutex`?

patrickangle replied:
`receivedAutomaticInspectionConfiguration` is only called from places where a lock is already taken.

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -743,9 +762,8 @@ static bool globalAutomaticInspectionState()
-    ASSERT(targetIdentifier == m_automaticInspectionCandidateTargetIdentifier);
-    if (targetIdentifier == m_automaticInspectionCandidateTargetIdentifier)
-        m_automaticInspectionPaused = false;
+    ASSERT(m_pausedAutomaticInspectionCandidates.contains(targetIdentifier));
+    m_pausedAutomaticInspectionCandidates.remove(targetIdentifier);
```
Does this need to be guarded by `m_mutex`?

patrickangle replied:
`receivedAutomaticInspectionRejectMessage` is only called from places where a lock is already taken.

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -99,21 +99,19 @@ static bool globalAutomaticInspectionState()
             }
+
+            shared.get().setPendingMainThreadInitialization(true);
+            if ([NSThread isMainThread])
```
NIT: should we use `WTF::isMainThread()` instead?  or maybe even `WTF::ensureOnMainThread`?

patrickangle replied:
Unfortunately in this case WTF needs `pthread_main_np` to return non-zero when we call `WTF::initialize()`, but both of your suggestions count the WebThread as being a main thread. Really, I guess to be fully accurate we should check `pthread_main_np` tho.

JosephPecoraro replied:
Patrick is correct, `WTF::isMainThread` allows for other threads and isn't appropriate to call `WTF::initialize`. Technically `WTF::isUIThread` is what we'd want, but I believe that requires initialization. Perhaps we could have a `WTF::isThreadAppropriateToInitialize` which I believe boils down to just the real main thread.

`-[NSThread isMainThread]` should just boil down to being a `pthread_main_np` call.

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -125,71 +123,95 @@ static bool globalAutomaticInspectionState()
+    auto singletonShouldImmediatelyInitialize = isMainThread();
+
+    {
+        Locker locker { m_mutex };
+        singletonShouldImmediatelyInitialize &= m_pendingMainThreadInitialization;
```
NIT: as a minor optimization, if we're not on the main thread then it seems like we could skip this, right?

dcrousso commented on Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorGlib.cpp:
```diff
@@ -94,7 +94,6 @@ void RemoteInspector::stopInternal(StopSource)
 
     updateHasActiveDebugSession();
 
-    m_automaticInspectionPaused = false;
```
does this need to `m_pausedAutomaticInspectionCandidates.clear()`?

patrickangle replied:
There isn't currently a way for glib (or the socket impl below) to ever add stuff to this list, but for completeness it can't hurt.

dcrousso commented on Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocket.cpp:
```diff
@@ -111,8 +111,6 @@ void RemoteInspector::stopInternal(StopSource)
 
     updateHasActiveDebugSession();
-
-    m_automaticInspectionPaused = false;
```
does this need to `m_pausedAutomaticInspectionCandidates.clear()`?

dcrousso commented on Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
```diff
@@ -268,6 +271,8 @@ class JS_EXPORT_PRIVATE RemoteInspector final
     RefPtr<RemoteInspectorXPCConnection> m_relayConnection;
     bool m_shouldReconnectToRelayOnFailure { false };
+
+    bool m_pendingMainThreadInitialization { false };
```
hmm should this also be `WTF_GUARDED_BY_LOCK(m_mutex)` (it seems like every use already does)?

patrickangle replied:
Sigh. It looks like I got the wires crossed in my brain and made `m_automaticInspectionEnabled` atomic instead of this, which I meant to 🤦🏻

dcrousso replied:
aha lol it happens :P

might be worth just making them both `WTF_GUARDED_BY_LOCK(m_mutex)` instead

patrickangle committed:
Web Inspector: [Cocoa] `RemoteInspector` fails to send automatic inspection candidates created on a non-main thread
https://bugs.webkit.org/show_bug.cgi?id=242267
rdar://51415024

Reviewed by Devin Rousso.

Automatic inspection of JSContexts works by spinning the thread a JSContext is created on when the feature is enabled at
the device level (via webinspectord) until we can ask an attached debugger(s) if we should automatically inspect that
context. At that same time, the attached debugger will let us know if we should formally pause script execution, or if
we should allow execution to continue normally. Upon receiving the yes/no from the debugger, we unspin the thread and if
requested start the debug session/pause JS execution.

This previously only reliably worked for the main thread and for contexts created after a JSContext was created on the
main thread. This was because there is initialization that must be done on the main thread (WTF::initialize). The issues
arose because while we would schedule initialization to later occur on the main thread, non-main-threads would still
carry on checking to see if, among other things, automatic inspection was enabled. Because the device-global automatic
inspection enablement wasn't read until after main thread initialization, we always declined to automatically inspect
non-main-thread contexts created before initialization had completed on the main thread.

This tangle of dependancies is unravelled by doing a few different things in tandem:
1. Get the global automatic inspection state before scheduling initialization - `notify_get_state` is designed for
cross-process and cross-thread messaging, so there is no need for us to be on the main thread to do so.
2. Keep track of the RemoteInspector singleton pending initialization.
3. Consider a "targets" (called debuggables by the client) for automatic inspection and spin them if automatic
inspection is enabled (since we now get that state before this point for all threads) and only bail if we are both
disabled and not pending initialization.

As part of this, it is now much more likely that multiple JSContexts may need to be considered candidates for automatic
inspection at the same time, so we also fix an existing limitation by allowing ourselves to keep track of multiple
candidates, since we may build a small set of candidates before initialization has completed.

* Source/JavaScriptCore/inspector/remote/RemoteInspector.cpp:
(Inspector::RemoteInspector::setupFailed):
(Inspector::RemoteInspector::setupCompleted):
(Inspector::RemoteInspector::waitingForAutomaticInspection):
* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
(Inspector::RemoteInspector::singleton):
(Inspector::RemoteInspector::initialize):
(Inspector::RemoteInspector::setPendingMainThreadInitialization):
(Inspector::RemoteInspector::updateAutomaticInspectionCandidate):
(Inspector::RemoteInspector::sendAutomaticInspectionCandidateMessage):
(Inspector::RemoteInspector::start):
(Inspector::RemoteInspector::stopInternal):
(Inspector::RemoteInspector::setupXPCConnectionIfNeeded):
(Inspector::RemoteInspector::xpcConnectionFailed):
(Inspector::RemoteInspector::receivedSetupMessage):
(Inspector::RemoteInspector::receivedAutomaticInspectionConfigurationMessage):
(Inspector::RemoteInspector::receivedAutomaticInspectionRejectMessage):
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorGlib.cpp:
(Inspector::RemoteInspector::stopInternal):
(Inspector::RemoteInspector::sendAutomaticInspectionCandidateMessage):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocket.cpp:
(Inspector::RemoteInspector::stopInternal):
(Inspector::RemoteInspector::sendAutomaticInspectionCandidateMessage):

Canonical link: https://commits.webkit.org/252569@main
